File: backend/omr_engine.py
"""
Core OMR detection engine calibrated for the NMMS-style answer sheet
(2 blocks of 20 questions x 4 options, circular bubbles, table-ruled).

Pipeline:
1. Load + auto-orient (EXIF) + resize
2. Locate the ruled answer-grid (largest line-grid contour)
3. Detect all bubble circles inside the grid (contour + circularity filter)
4. Cluster circle centers into a canonical 20-row x 8-column bubble grid
   (robust to missing/dashed lines - anchors on the bubbles themselves)
5. For each of the 160 bubble cells, compute a multi-metric fill score
6. Resolve each question -> answer option (1-4), flag blank/multiple
"""
import cv2
import numpy as np
from sklearn.cluster import KMeans
from dataclasses import dataclass, field
from typing import Optional


@dataclass
class BubbleResult:
    question: int
    selected: Optional[int]   # 1-4, None if blank
    is_multiple: bool
    multiple_options: list
    confidences: dict         # option -> fill confidence %
    flagged: bool = False
    flag_reason: str = ""


@dataclass
class ScanResult:
    answers: dict              # question -> BubbleResult
    quality: dict               # blur, brightness, etc.
    debug_image: Optional[np.ndarray] = None
    warnings: list = field(default_factory=list)


def load_image_exif_safe(path_or_bytes):
    """Load image respecting EXIF orientation."""
    if isinstance(path_or_bytes, (bytes, bytearray)):
        arr = np.frombuffer(path_or_bytes, np.uint8)
        img = cv2.imdecode(arr, cv2.IMREAD_COLOR)
    else:
        img = cv2.imread(path_or_bytes, cv2.IMREAD_COLOR)
    if img is None:
        raise ValueError("Could not decode image")
    return img


def blur_score(gray):
    return cv2.Laplacian(gray, cv2.CV_64F).var()


def resize_max(img, max_dim=1600):
    h, w = img.shape[:2]
    scale = max_dim / max(h, w)
    if scale < 1:
        img = cv2.resize(img, (int(w * scale), int(h * scale)))
    return img


def find_answer_grid_bbox(gray):
    """Find the bounding box of the ruled answer table using morphological
    horizontal/vertical line extraction."""
    h, w = gray.shape
    blur = cv2.GaussianBlur(gray, (5, 5), 0)
    th = cv2.adaptiveThreshold(blur, 255, cv2.ADAPTIVE_THRESH_GAUSSIAN_C,
                                cv2.THRESH_BINARY_INV, 25, 10)
    hk = cv2.getStructuringElement(cv2.MORPH_RECT, (max(20, w // 30), 1))
    horiz = cv2.dilate(cv2.erode(th, hk), hk)
    vk = cv2.getStructuringElement(cv2.MORPH_RECT, (1, max(20, h // 40)))
    vert = cv2.dilate(cv2.erode(th, vk), vk)
    grid = cv2.add(horiz, vert)
    contours, _ = cv2.findContours(grid, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
    if not contours:
        return None
    contours = sorted(contours, key=cv2.contourArea, reverse=True)
    c = contours[0]
    x, y, ww, hh = cv2.boundingRect(c)
    # sanity: grid should be a big chunk of the image
    if ww * hh < 0.15 * w * h:
        return None
    return x, y, ww, hh


def detect_circles(roi_gray):
    """Detect circular bubble candidates within the answer-grid ROI."""
    blur = cv2.GaussianBlur(roi_gray, (5, 5), 0)
    th = cv2.adaptiveThreshold(blur, 255, cv2.ADAPTIVE_THRESH_GAUSSIAN_C,
                                cv2.THRESH_BINARY_INV, 21, 8)
    cnts, _ = cv2.findContours(th, cv2.RETR_LIST, cv2.CHAIN_APPROX_SIMPLE)
    h, w = roi_gray.shape
    # expected bubble radius: table is divided into ~10 cols x 21 rows
    approx_cell = min(w / 10, h / 21)
    r_lo, r_hi = approx_cell * 0.18, approx_cell * 0.42

    raw = []
    for c in cnts:
        area = cv2.contourArea(c)
        if area < 30:
            continue
        (cx, cy), radius = cv2.minEnclosingCircle(c)
        if radius < r_lo or radius > r_hi:
            continue
        circularity = area / (np.pi * radius * radius + 1e-5)
        if circularity < 0.45:
            continue
        raw.append((cx, cy, radius))

    # non-max suppression: merge near-duplicate detections (outer/inner contours)
    raw.sort(key=lambda t: -t[2])
    kept = []
    for cx, cy, r in raw:
        dup = False
        for kx, ky, kr in kept:
            if (cx - kx) ** 2 + (cy - ky) ** 2 < (max(r, kr) * 0.9) ** 2:
                dup = True
                break
        if not dup:
            kept.append((cx, cy, r))
    return kept


def cluster_grid(circles, n_rows=20, n_cols=8):
    """Cluster raw circle detections into a canonical row x col grid.
    Returns dict[(row,col)] -> (cx,cy,r) and calibration diagnostics."""
    if len(circles) < n_rows * n_cols * 0.5:
        raise ValueError(
            f"Only {len(circles)} bubble candidates found; expected ~{n_rows*n_cols}. "
            "Image quality too low for reliable detection."
        )

    pts = np.array([[c[0], c[1]] for c in circles])

    # cluster columns first (8 option-columns across 2 blocks)
    col_km = KMeans(n_clusters=n_cols, n_init=10, random_state=0).fit(pts[:, 0:1])
    col_centers = col_km.cluster_centers_.flatten()
    col_order = np.argsort(col_centers)
    col_rank = {old: new for new, old in enumerate(col_order)}
    col_labels = np.array([col_rank[l] for l in col_km.labels_])

    # cluster rows (20 question rows) using y only
    row_km = KMeans(n_clusters=n_rows, n_init=10, random_state=0).fit(pts[:, 1:2])
    row_centers = row_km.cluster_centers_.flatten()
    row_order = np.argsort(row_centers)
    row_rank = {old: new for new, old in enumerate(row_order)}
    row_labels = np.array([row_rank[l] for l in row_km.labels_])

    grid = {}
    for i, (cx, cy, r) in enumerate(circles):
        key = (row_labels[i], col_labels[i])
        # if collision keep the more "circular"/plausible one (first wins is fine
        # since duplicates already suppressed)
        if key not in grid:
            grid[key] = (cx, cy, r)

    return grid, sorted(row_centers[row_order]), sorted(col_centers[col_order])


def fill_metrics(gray_roi, cx, cy, r):
    """Compute fill confidence for a single bubble using multiple metrics."""
    h, w = gray_roi.shape
    x0, x1 = max(0, int(cx - r)), min(w, int(cx + r))
    y0, y1 = max(0, int(cy - r)), min(h, int(cy + r))
    patch = gray_roi[y0:y1, x0:x1]
    if patch.size == 0:
        return 0.0, 0.0

    mask = np.zeros(patch.shape, dtype=np.uint8)
    pcx, pcy = cx - x0, cy - y0
    inner_r = max(1, int(r * 0.72))  # avoid counting the printed ring itself
    cv2.circle(mask, (int(pcx), int(pcy)), inner_r, 255, -1)

    inner_pixels = patch[mask == 255]
    if inner_pixels.size == 0:
        return 0.0, 0.0

    mean_intensity = float(inner_pixels.mean())
    # local threshold: compare against the bubble's own local background sample
    ring_mask = np.zeros(patch.shape, dtype=np.uint8)
    cv2.circle(ring_mask, (int(pcx), int(pcy)), max(1, int(r * 0.95)), 255, -1)
    cv2.circle(ring_mask, (int(pcx), int(pcy)), inner_r, 0, -1)
    dark_ratio = float((inner_pixels < 150).sum()) / inner_pixels.size

    return mean_intensity, dark_ratio


def score_all_bubbles(gray_roi, grid, n_rows, n_cols):
    """Returns dict[(row,col)] -> confidence(0-100, higher = more filled)."""
    raw_scores = {}
    for (row, col), (cx, cy, r) in grid.items():
        mean_i, dark_ratio = fill_metrics(gray_roi, cx, cy, r)
        raw_scores[(row, col)] = (mean_i, dark_ratio)

    # normalize mean intensity per-row (handles lighting gradient across sheet)
    confidences = {}
    for row in range(n_rows):
        row_vals = [raw_scores[(row, c)][0] for c in range(n_cols) if (row, c) in raw_scores]
        if not row_vals:
            continue
        lo, hi = min(row_vals), max(row_vals)
        span = max(hi - lo, 1e-5)
        for col in range(n_cols):
            if (row, col) not in raw_scores:
                continue
            mean_i, dark_ratio = raw_scores[(row, col)]
            # darker (lower mean intensity) relative to row peers = more filled
            rel_fill = (hi - mean_i) / span
            score = 0.55 * rel_fill + 0.45 * dark_ratio
            confidences[(row, col)] = round(float(score) * 100, 1)
    return confidences


FILL_THRESHOLD = 42.0       # confidence % above which a bubble counts as marked
AMBIGUOUS_MARGIN = 12.0     # if 2nd place is within this margin of 1st -> multi-mark flag


def resolve_answers(confidences, n_rows=20, n_cols=8, block_size=4):
    """Convert per-bubble confidences into per-question answers.
    Columns 0-3 = block A (Q1-20), columns 4-7 = block B (Q21-40)."""
    results = {}
    for row in range(n_rows):
        for block, col_offset, q_offset in [(0, 0, 0), (1, 4, 20)]:
            qnum = row + 1 + q_offset
            opt_scores = {}
            for opt in range(4):
                col = col_offset + opt
                conf = confidences.get((row, col))
                if conf is not None:
                    opt_scores[opt + 1] = conf
            if not opt_scores:
                results[qnum] = BubbleResult(qnum, None, False, [], {}, True, "not detected")
                continue
            marked = [o for o, c in opt_scores.items() if c >= FILL_THRESHOLD]
            marked.sort(key=lambda o: -opt_scores[o])

            if len(marked) == 0:
                results[qnum] = BubbleResult(qnum, None, False, [], opt_scores, False, "blank")
            elif len(marked) == 1:
                results[qnum] = BubbleResult(qnum, marked[0], False, [], opt_scores)
            else:
                top, second = marked[0], marked[1]
                if opt_scores[top] - opt_scores[second] < AMBIGUOUS_MARGIN:
                    results[qnum] = BubbleResult(
                        qnum, None, True, marked, opt_scores, True, "multiple marks"
                    )
                else:
                    results[qnum] = BubbleResult(qnum, top, False, [], opt_scores)
    return results


def scan_omr_sheet(image_bytes_or_path, n_rows=20, n_cols=8, debug=False) -> ScanResult:
    img = load_image_exif_safe(image_bytes_or_path)
    img = resize_max(img, 1600)
    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

    warnings = []
    bscore = blur_score(gray)
    brightness = float(gray.mean())
    if bscore < 60:
        warnings.append(f"Image may be blurry (sharpness score {bscore:.0f}, recommend >60)")
    if brightness < 60:
        warnings.append("Image is quite dark; consider better lighting")
    if brightness > 220:
        warnings.append("Image is overexposed; consider reducing glare")

    bbox = find_answer_grid_bbox(gray)
    if bbox is None:
        raise ValueError("Could not locate the answer grid in this image. "
                          "Ensure the full sheet with its ruled table is visible.")
    x, y, w, h = bbox
    roi_gray = gray[y:y + h, x:x + w]

    circles = detect_circles(roi_gray)
    grid, row_centers, col_centers = cluster_grid(circles, n_rows, n_cols)

    confidences = score_all_bubbles(roi_gray, grid, n_rows, n_cols)
    answers = resolve_answers(confidences, n_rows, n_cols)

    debug_img = None
    if debug:
        debug_img = img.copy()
        for (row, col), (cx, cy, r) in grid.items():
            conf = confidences.get((row, col), 0)
            color = (0, 0, 255) if conf >= FILL_THRESHOLD else (0, 200, 0)
            cv2.circle(debug_img, (int(cx + x), int(cy + y)), int(r), color, 2)

    return ScanResult(answers=answers, quality={
        "blur_score": round(bscore, 1),
        "brightness": round(brightness, 1),
        "bubbles_detected": len(circles),
    }, debug_image=debug_img, warnings=warnings)

File: android/app/src/main/java/com/omr/capture/OmrEngine.kt
package com.omr.capture

import org.opencv.core.Core
import org.opencv.core.CvType
import org.opencv.core.Mat
import org.opencv.core.MatOfPoint
import org.opencv.core.MatOfPoint2f
import org.opencv.core.Point
import org.opencv.core.Rect
import org.opencv.core.Scalar
import org.opencv.core.Size
import org.opencv.imgproc.CLAHE
import org.opencv.imgproc.Imgproc
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

/**
 * On-device port of `omr_engine.py` (+ the post-capture cleanup from
 * `frame_quality.process_captured_image`). Detects bubble circles inside the
 * ruled answer grid, clusters them into a 20x8 grid with a lightweight 1-D
 * k-means, scores each bubble's fill and resolves per-question answers.
 *
 * Calibrated for the NMMS sheet: 40 questions in 2 blocks of 20, 4 options each.
 */
object OmrEngine {

    private const val N_ROWS = 20
    private const val N_COLS = 8
    private const val FILL_THRESHOLD = 42.0
    private const val AMBIGUOUS_MARGIN = 12.0

    /** Per-question detection result. selected is 1-4 or null when blank. */
    data class BubbleResult(
        val question: Int,
        val selected: Int?,
        val isMultiple: Boolean,
        val flagged: Boolean = false,
        val flagReason: String = "",
    )

    data class ScanResult(
        val answers: Map<Int, BubbleResult>,
        val bubblesDetected: Int,
        val blurScore: Double,
        val brightness: Double,
        val warnings: List<String>,
    )

    // ---------------------------------------------------------------------
    // Post-capture cleanup: perspective correction + deskew, shadow removal
    // (background division), contrast enhancement (CLAHE).
    // ---------------------------------------------------------------------
    fun process(bgr: Mat): Mat {
        val gray = Mat()
        Imgproc.cvtColor(bgr, gray, Imgproc.COLOR_BGR2GRAY)
        val corners = FrameQuality.detectGridQuad(gray)

        var work = bgr
        if (corners != null) {
            val (tl, tr, br, bl) = corners
            val gw = max(hyp(tr, tl), hyp(br, bl)).roundToInt()
            val gh = max(hyp(bl, tl), hyp(br, tr)).roundToInt()
            val margin = (0.04 * max(gw, gh)).roundToInt()
            val src = MatOfPoint2f(*corners)
            val dst = MatOfPoint2f(
                Point(margin.toDouble(), margin.toDouble()),
                Point((margin + gw).toDouble(), margin.toDouble()),
                Point((margin + gw).toDouble(), (margin + gh).toDouble()),
                Point(margin.toDouble(), (margin + gh).toDouble()),
            )
            val m = Imgproc.getPerspectiveTransform(src, dst)
            val warped = Mat()
            Imgproc.warpPerspective(
                bgr, warped, m, Size((gw + 2 * margin).toDouble(), (gh + 2 * margin).toDouble()),
                Imgproc.INTER_CUBIC, Core.BORDER_REPLICATE, Scalar(0.0)
            )
            work = warped
        }

        val g = Mat()
        Imgproc.cvtColor(work, g, Imgproc.COLOR_BGR2GRAY)
        val dil = Mat()
        Imgproc.dilate(g, dil, Mat.ones(Size(7.0, 7.0), CvType.CV_8U))
        val bg = Mat()
        Imgproc.medianBlur(dil, bg, 31)
        val norm = Mat()
        Core.divide(g, bg, norm, 255.0)

        val clahe: CLAHE = Imgproc.createCLAHE(2.0, Size(8.0, 8.0))
        val enhanced = Mat()
        clahe.apply(norm, enhanced)

        val out = Mat()
        Imgproc.cvtColor(enhanced, out, Imgproc.COLOR_GRAY2BGR)
        return out
    }

    private fun hyp(a: Point, b: Point) = Math.hypot(a.x - b.x, a.y - b.y)

    // ---------------------------------------------------------------------
    // Scan: locate grid bbox, detect circles, cluster, score, resolve.
    // ---------------------------------------------------------------------
    fun scan(bgr: Mat): ScanResult {
        val img = resizeMax(bgr, 1600)
        val gray = Mat()
        Imgproc.cvtColor(img, gray, Imgproc.COLOR_BGR2GRAY)

        val warnings = ArrayList<String>()
        val bscore = laplacianVar(gray)
        val brightness = Core.mean(gray).`val`[0]
        if (bscore < 60) warnings.add("Image may be blurry (sharpness ${bscore.roundToInt()}, recommend >60)")
        if (brightness < 60) warnings.add("Image is quite dark; consider better lighting")
        if (brightness > 220) warnings.add("Image is overexposed; consider reducing glare")

        val bbox = findAnswerGridBbox(gray)
            ?: throw IllegalStateException("Could not locate the answer grid. Ensure the full ruled table is visible.")
        val roi = gray.submat(bbox)

        val circles = detectCircles(roi)
        if (circles.size < N_ROWS * N_COLS * 0.5) {
            throw IllegalStateException("Only ${circles.size} bubbles found; expected ~${N_ROWS * N_COLS}. Image quality too low.")
        }

        val grid = clusterGrid(circles)
        val confidences = scoreAllBubbles(roi, grid)
        val answers = resolveAnswers(confidences)

        return ScanResult(answers, circles.size, round1(bscore), round1(brightness), warnings)
    }

    private data class Circle(val cx: Double, val cy: Double, val r: Double)

    private fun findAnswerGridBbox(gray: Mat): Rect? {
        val w = gray.cols(); val h = gray.rows()
        val blur = Mat(); Imgproc.GaussianBlur(gray, blur, Size(5.0, 5.0), 0.0)
        val th = Mat()
        Imgproc.adaptiveThreshold(blur, th, 255.0, Imgproc.ADAPTIVE_THRESH_GAUSSIAN_C, Imgproc.THRESH_BINARY_INV, 25, 10.0)
        val hk = Imgproc.getStructuringElement(Imgproc.MORPH_RECT, Size(max(20, w / 30).toDouble(), 1.0))
        val horiz = Mat(); Imgproc.erode(th, horiz, hk); Imgproc.dilate(horiz, horiz, hk)
        val vk = Imgproc.getStructuringElement(Imgproc.MORPH_RECT, Size(1.0, max(20, h / 40).toDouble()))
        val vert = Mat(); Imgproc.erode(th, vert, vk); Imgproc.dilate(vert, vert, vk)
        val grid = Mat(); Core.add(horiz, vert, grid)
        val contours = ArrayList<MatOfPoint>()
        Imgproc.findContours(grid, contours, Mat(), Imgproc.RETR_EXTERNAL, Imgproc.CHAIN_APPROX_SIMPLE)
        if (contours.isEmpty()) return null
        val c = contours.maxByOrNull { Imgproc.contourArea(it) }!!
        val r = Imgproc.boundingRect(c)
        if (r.width.toDouble() * r.height < 0.15 * w * h) return null
        return r
    }

    private fun detectCircles(roi: Mat): List<Circle> {
        val blur = Mat(); Imgproc.GaussianBlur(roi, blur, Size(5.0, 5.0), 0.0)
        val th = Mat()
        Imgproc.adaptiveThreshold(blur, th, 255.0, Imgproc.ADAPTIVE_THRESH_GAUSSIAN_C, Imgproc.THRESH_BINARY_INV, 21, 8.0)
        val cnts = ArrayList<MatOfPoint>()
        Imgproc.findContours(th, cnts, Mat(), Imgproc.RETR_LIST, Imgproc.CHAIN_APPROX_SIMPLE)
        val approxCell = min(roi.cols() / 10.0, roi.rows() / 21.0)
        val rLo = approxCell * 0.18; val rHi = approxCell * 0.42

        val raw = ArrayList<Circle>()
        for (c in cnts) {
            val area = Imgproc.contourArea(c)
            if (area < 30) continue
            val center = Point(); val radius = FloatArray(1)
            Imgproc.minEnclosingCircle(MatOfPoint2f(*c.toArray()), center, radius)
            val rad = radius[0].toDouble()
            if (rad < rLo || rad > rHi) continue
            val circ = area / (Math.PI * rad * rad + 1e-5)
            if (circ < 0.45) continue
            raw.add(Circle(center.x, center.y, rad))
        }

        // non-max suppression: merge near-duplicate detections
        raw.sortByDescending { it.r }
        val kept = ArrayList<Circle>()
        for (c in raw) {
            var dup = false
            for (k in kept) {
                val d2 = (c.cx - k.cx) * (c.cx - k.cx) + (c.cy - k.cy) * (c.cy - k.cy)
                if (d2 < (max(c.r, k.r) * 0.9) * (max(c.r, k.r) * 0.9)) { dup = true; break }
            }
            if (!dup) kept.add(c)
        }
        return kept
    }

    private fun clusterGrid(circles: List<Circle>): Map<Pair<Int, Int>, Circle> {
        val colLabels = kmeans1d(circles.map { it.cx }, N_COLS)
        val rowLabels = kmeans1d(circles.map { it.cy }, N_ROWS)
        val grid = HashMap<Pair<Int, Int>, Circle>()
        for (i in circles.indices) {
            val key = Pair(rowLabels[i], colLabels[i])
            if (key !in grid) grid[key] = circles[i]
        }
        return grid
    }

    /** Lloyd's 1-D k-means; returns rank-ordered cluster labels (0..k-1 by value). */
    private fun kmeans1d(values: List<Double>, k: Int, iters: Int = 25): IntArray {
        val n = values.size
        val sorted = values.sorted()
        // init centers evenly across the sorted range
        val centers = DoubleArray(k) { sorted[((it + 0.5) / k * (n - 1)).toInt().coerceIn(0, n - 1)] }
        val labels = IntArray(n)
        repeat(iters) {
            for (i in 0 until n) {
                var best = 0; var bestD = Double.MAX_VALUE
                for (c in 0 until k) {
                    val d = Math.abs(values[i] - centers[c])
                    if (d < bestD) { bestD = d; best = c }
                }
                labels[i] = best
            }
            val sum = DoubleArray(k); val cnt = IntArray(k)
            for (i in 0 until n) { sum[labels[i]] += values[i]; cnt[labels[i]]++ }
            for (c in 0 until k) if (cnt[c] > 0) centers[c] = sum[c] / cnt[c]
        }
        // rank centers so label 0 = smallest coordinate, matching the Python order
        val order = (0 until k).sortedBy { centers[it] }
        val rank = IntArray(k)
        for (newIdx in order.indices) rank[order[newIdx]] = newIdx
        return IntArray(n) { rank[labels[it]] }
    }

    /** dark-fill confidence 0-100 per bubble, normalized per-row for lighting. */
    private fun scoreAllBubbles(roi: Mat, grid: Map<Pair<Int, Int>, Circle>): Map<Pair<Int, Int>, Double> {
        val raw = HashMap<Pair<Int, Int>, Pair<Double, Double>>()
        for ((key, c) in grid) raw[key] = fillMetrics(roi, c)

        val confidences = HashMap<Pair<Int, Int>, Double>()
        for (row in 0 until N_ROWS) {
            val rowVals = (0 until N_COLS).mapNotNull { raw[Pair(row, it)]?.first }
            if (rowVals.isEmpty()) continue
            val lo = rowVals.min(); val hi = rowVals.max()
            val span = max(hi - lo, 1e-5)
            for (col in 0 until N_COLS) {
                val m = raw[Pair(row, col)] ?: continue
                val relFill = (hi - m.first) / span
                val score = 0.55 * relFill + 0.45 * m.second
                confidences[Pair(row, col)] = round1(score * 100)
            }
        }
        return confidences
    }

    /** returns (meanIntensity, darkRatio) inside the bubble's inner disc. */
    private fun fillMetrics(roi: Mat, c: Circle): Pair<Double, Double> {
        val w = roi.cols(); val h = roi.rows()
        val x0 = max(0, (c.cx - c.r).toInt()); val x1 = min(w, (c.cx + c.r).toInt())
        val y0 = max(0, (c.cy - c.r).toInt()); val y1 = min(h, (c.cy + c.r).toInt())
        if (x1 <= x0 || y1 <= y0) return Pair(255.0, 0.0)
        val patch = roi.submat(Rect(x0, y0, x1 - x0, y1 - y0))

        val mask = Mat.zeros(patch.size(), CvType.CV_8UC1)
        val pcx = c.cx - x0; val pcy = c.cy - y0
        val innerR = max(1, (c.r * 0.72).toInt())
        Imgproc.circle(mask, Point(pcx, pcy), innerR, Scalar(255.0), -1)

        val meanIntensity = Core.mean(patch, mask).`val`[0]

        val dark = Mat()
        Imgproc.threshold(patch, dark, 149.0, 255.0, Imgproc.THRESH_BINARY_INV)
        Core.bitwise_and(dark, mask, dark)
        val innerCount = Core.countNonZero(mask).coerceAtLeast(1)
        val darkRatio = Core.countNonZero(dark).toDouble() / innerCount
        return Pair(meanIntensity, darkRatio)
    }

    private fun resolveAnswers(conf: Map<Pair<Int, Int>, Double>): Map<Int, BubbleResult> {
        val results = HashMap<Int, BubbleResult>()
        for (row in 0 until N_ROWS) {
            for ((colOffset, qOffset) in listOf(Pair(0, 0), Pair(4, 20))) {
                val qnum = row + 1 + qOffset
                val optScores = HashMap<Int, Double>()
                for (opt in 0 until 4) {
                    conf[Pair(row, colOffset + opt)]?.let { optScores[opt + 1] = it }
                }
                if (optScores.isEmpty()) {
                    results[qnum] = BubbleResult(qnum, null, false, true, "not detected"); continue
                }
                val marked = optScores.filter { it.value >= FILL_THRESHOLD }.keys
                    .sortedByDescending { optScores[it] }
                results[qnum] = when {
                    marked.isEmpty() -> BubbleResult(qnum, null, false, false, "blank")
                    marked.size == 1 -> BubbleResult(qnum, marked[0], false)
                    else -> {
                        val top = marked[0]; val second = marked[1]
                        if (optScores[top]!! - optScores[second]!! < AMBIGUOUS_MARGIN)
                            BubbleResult(qnum, null, true, true, "multiple marks")
                        else BubbleResult(qnum, top, false)
                    }
                }
            }
        }
        return results
    }

    private fun resizeMax(img: Mat, maxDim: Int): Mat {
        val scale = maxDim.toDouble() / max(img.rows(), img.cols())
        if (scale >= 1) return img
        val out = Mat()
        Imgproc.resize(img, out, Size(img.cols() * scale, img.rows() * scale))
        return out
    }

    private fun laplacianVar(gray: Mat): Double {
        val lap = Mat(); Imgproc.Laplacian(gray, lap, CvType.CV_64F)
        val std = MatOfDoubleWrap()
        Core.meanStdDev(lap, MatOfDoubleWrap().mat, std.mat)
        val s = std.value(); return s * s
    }

    private fun round1(v: Double) = Math.round(v * 10.0) / 10.0
}
